### PR TITLE
gh-99029: Fix handling of `PureWindowsPath('C:\<blah>').relative_to('C:')`

### PR DESCRIPTION
**Pull request by barneygale**
`relative_to()` now treats naked drive paths as relative. This brings its behaviour in line with other parts of pathlib, ~~and with `ntpath.relpath()`, and so allows us to factor out the pathlib-specific implementation.~~

* Issue: gh-99029

### CODE REVIEW

**Comment by domragusa** [comment]
The implementation of os.path.relpath accesses the filesystem making the two paths absolute and then calculating the common path, using that function to implement a PurePath method is wrong becase it should work without the any file or directory being there (its definition is "Base class for manipulating paths without I/O"), the use case that it breaks is when we don't have files on disk for example I use it to create the structure of a zip archive.

Removing tests is not a good idea.

Mixing relative and absolute paths should raise an exception as you stated and should be fixed.

**Comment by barneygale** [comment]
I didn't remove any tests. Please share an example of where this breaks.

**Comment by barneygale** [comment]
I told a lie - I actually did remove two assertions that supply strings to the method. Will fix!

**Comment by domragusa** [comment]
I thought that os.path.relpath accessed the filesystem, but taking a better look at it I guess it doesn't.

I don't have a Windows machine and I can't provide a concrete example, I'm sorry, anyway it calls os.path.abspath and following the chain we arrive at a Windows api call [GetFullPathNameW](https://learn.microsoft.com/en-us/windows/win32/api/fileapi/nf-fileapi-getfullpathnamew), that's where I expect PurePath to break because it should be platform independent and this change would make it depend on the result of a winapi call.
Can you test it with names and characters that are reserved on Windows but are allowed on Posix-systems?

I would have just tweaked the fail condition to make sure that we raise the exception for the bug you noticed.

**Comment by barneygale** [comment]
:+1: thanks, I'll try that out!

It's true that `relpath()` can call `os.getcwd()`. However, the working directory only contributes to the result _if we pass one relative path and one absolute_. And that's specifically disallowed in pathlib's `relative_to()`. In other cases, the two matching prefixes cancel eachother out in `commonpath()`.

Therefore, the fact that `relpath()` calls `abspath()` when given two relative paths is regrettable, and probably worth fixing. But it shouldn't have an observable effect here.

**Comment by domragusa** [comment]
It's not just getcwd, for example "c:" is a valid filename on Linux: on Windows after abspath you would get the current directory on the C drive while on Linux it's the file named "c:" in the current working directory so the calculations that follow would be wrong, there are also reserved names on Windows like COM or CON and characters that are not allowed.

**Comment by barneygale** [comment]
I don't think that matters - `os.getcwd()` can return something totally nonsensical and we wouldn't care, as long as it returns the same thing twice in a row, so that the shared prefix can be eliminated by `commonpath()`. Right?

**Comment by domragusa** [comment]
You're referencing `os.getcwd()` because you're looking at the implementation of `os.path.abspath` in `os.posixpath` but on Windows `os.path` resolves to `os.ntpath` and in that module `os.path.abspath` calls `_getfullpathname()` that calls the windows api.

**Comment by barneygale** [comment]
Again, I don't think it matters. If you think it does, could you please provide a reproduction case. Thank you.

**Comment by barneygale** [comment]
I've logged #99199 to cover `relpath()` needlessly calling `abspath()` when the paths' anchors match.

**Comment by barneygale** [comment]
Turns out there's another incompatibility related to leading `../` segments - see https://github.com/python/cpython/issues/99199#issuecomment-1312374296

I've revised the implementation to not use `relpath()` :)

File: Lib/pathlib.py [hunk 1]
```diff
@@ -634,57 +634,27 @@ def relative_to(self, *other, walk_up=False):
...
-            error_message = "{!r} is not in the subpath of {!r}"
-            up_parts = []
-        error_message += " OR one path is relative and the other is absolute."
-        if failure:
-            formatted = self._format_parsed_parts(other_drv, other_root, other_parts)
-            raise ValueError(error_message.format(str(self), str(formatted)))
-        path_parts = up_parts + abs_parts[num_common_parts:]
-        new_root = root if num_common_parts == 1 else ''
-        return self._from_parsed_parts('', new_root, path_parts)
+            raise ValueError("{!r} and {!r} have different anchors".format(str(self), str(other)))
```
Comment (brettcannon) [inline Lib/pathlib.py]:
```suggestion
            raise ValueError(f"{self!r} and {other!r} have different anchors")
```

File: Lib/pathlib.py [hunk 2]
```diff
@@ -634,57 +634,27 @@ def relative_to(self, *other, walk_up=False):
...
+        if step and not walk_up:
+            raise ValueError("{!r} is not in the subpath of {!r}".format(str(self), str(other)))
```
Comment (brettcannon) [inline Lib/pathlib.py]:
```suggestion
            raise ValueError(f"{self!r} is not in the subpath of {other!r}")
```

**Review by brettcannon — CHANGES_REQUESTED** [review]
I believe this is in line with the result of the discussion in the issue, so only last change is to use f-strings!

File: Lib/pathlib.py (hunk 1 above)
Comment (barneygale) [reply Lib/pathlib.py]: This isn't equivalent though, as it's missing the `str()` step! Could do `{str(self)!r}`?

**Comment by barneygale** [comment]
I have made the requested changes; please review again.

**Review by brettcannon — APPROVED** [review]

**Comment by brettcannon** [comment]
Thanks!

**Comment by barneygale** [comment]
Thank you!

**Pull request merged**